src/auto_scheduler/compute_dag.cc

```diff
@@ -665,9 +666,349 @@ ComputeDAG::ComputeDAG(Array<te::Tensor> tensors) {
+          if (handled_ops.count(placeholder_op)) {
+            continue;
+          }
+
+          // skip the op that is not direct consumer of this placeholder,
+          // mostly due to cache read/write.
+          bool direct_consumer = false;
+          for (auto& t : op->InputTensors()) {
+            if (t->op == placeholder_op) {
+              direct_consumer = true;
+              break;
+            }
+          }
+          if (!direct_consumer) {
+            continue;
+          }
+
+          std::set<std::string> placeholder_axis_names;
+          get_ori_layout(&placeholder_axis_names, op, placeholder);
+
+          Array<PrimExpr> new_shape;
+          std::string new_layout = get_new_layout(&new_shape, state, stage_id, stage, op,
+                                                  placeholder, placeholder_axis_names);
+
+          handled_ops.insert(placeholder_op);
+
+          Array<te::Operation> old_ops = pdag->ops;
+          ArrayNode* pops = pdag->ops.CopyOnWrite();
+
+          // Create new placeholder
+          te::Operation new_placeholder_op;
+          new_placeholder_op = te::PlaceholderOp(placeholder_op->name, new_shape,
+                                                 placeholder_op.as<te::PlaceholderOpNode>()->dtype);
+
+          te::Operation new_compute_op, old_compute_op;
+          Array<PrimExpr> new_body;
+          IndexRewriter index_rewriter(placeholder_op, new_layout);
+          for (auto& op : old_ops) {
+            if (auto* pop = op.as<te::ComputeOpNode>()) {
+              bool need_update = false;
+              for (auto& t : op->InputTensors()) {
+                if (t->op == placeholder_op) {
+                  need_update = true;
+                  break;
+                }
+              }
+              if (need_update) {
+                for (auto& body : pop->body) {
+                  new_body.push_back(index_rewriter.Rewrite(body));
+                }
+                old_compute_op = op;
+                CHECK(!new_compute_op.defined());
+                new_compute_op =
+                    te::ComputeOp(pop->name, pop->tag, pop->attrs, pop->axis, new_body);
+              }
+            }
+          }
+
+          // construct the map from old_op to new_op
+          std::unordered_map<te::Operation, te::Operation> updated_ops;
+          for (size_t i = 0; i < old_ops.size(); ++i) {
+            auto old_op = old_ops[i];
+            if (old_op == placeholder_op) {
+              pops->SetItem(i, new_placeholder_op);
+              updated_ops[placeholder_op] = new_placeholder_op;
+            } else if (old_op == old_compute_op) {
+              pops->SetItem(i, new_compute_op);
+              updated_ops[old_compute_op] = new_compute_op;
+            } else {
+              pops->SetItem(i, old_op);
+            }
+          }
+
+          // Because ops is sorted in topo-order, only do one pass linear scan here.
```

Comment:
Add comments talking about the purpose.
```suggestion
          // Because ops is sorted in topo-order, we only need one pass to (what).
```